### PR TITLE
Added customer store

### LINKED ISSUE #700: Adminpanel kundestore
- [x] exponere kunder, valgt kunde, prosjekter, valgt prosjekt og aktaktiviteter
- [x] oppdatere valgte kunde og prosjekt

sp: 13

### PR DESCRIPTION
**Pull request by yvindo**
- Added customer store with functions for selecting customer, project and activity
- Updated components to use data and functions from store

Closes #700

### CODE REVIEW

File: packages/adminpanel-v2/src/components/customerpage/customerdetails/activities/Activities.svelte
Comment (stormboat) [inline packages/adminpanel-v2/src/components/customerpage/customerdetails/activities/Activities.svelte:1]: What does the `$:` in general mean?

File: packages/adminpanel-v2/src/stores/CustomerStore.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	const fetchCustomers = () => {
+		//fetch from backend
+	}
+
+	const setCustomers = () => {
+		update((n) => {n.customers = mock; return n})
+		
+	}
+
+    const setCustomer = (id: number) => {
```
Comment (stormboat) [inline packages/adminpanel-v2/src/stores/CustomerStore.ts:37]: Can we after this PR is pushed, write to a temporary store and play with the variables during development?

Comment (stormboat) [inline packages/adminpanel-v2/src/stores/CustomerStore.ts:37]: Is this the customer_id (kunde_id) that Marcus talked about, or is this just a random generated ID number we use to have uniuqe customer IDs?

File: packages/adminpanel-v2/src/stores/CustomerStore.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		if (id != get(customerStore).active.customer) {
+			update((n) => {n.active.customer = id; return n})
+			update((n) => {n.active.project = undefined; return n})
+			update((n) => {n.active.activity = undefined; return n})
+		}
+	}
+
+	const setProject = (id: number) => {
```
Comment (stormboat) [inline packages/adminpanel-v2/src/stores/CustomerStore.ts:45]: Same as above, when we setProject, do we write to a temporary memory per now? And is the store a kind of temporary memory during dev? Or do you have to write to a database directly?

And when you write to store, is there any hard constraints on how much you can write?

File: packages/adminpanel-v2/src/stores/CustomerStore.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			update((n) => {n.active.project = id; return n})
+			update((n) => {n.active.activity = undefined; return n})
+		}
+	}
+
+	const setAvtivity = (id: number) => {
+		update((n) => {n.active.activity = id; return n})
+	}
+
+	const getActiveCustomer = () => {
```
Comment (stormboat) [inline packages/adminpanel-v2/src/stores/CustomerStore.ts:56]: This is the chosen customer which is chosen by the 'frontend'er right?

File: packages/adminpanel-v2/src/stores/CustomerStore.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	const getActiveActivity = () => {
+		const store = get(customerStore)
+		return store.customers.find((c) => c.id == store.active.customer)?.projects.find((p) => p.id == store.active.project)?.activities.find((a) => a.id == store.active.activity)
+	}
+
+
+	return {
+		subscribe,
+		setCustomers,
+        setCustomer,
```
Comment (stormboat) [inline packages/adminpanel-v2/src/stores/CustomerStore.ts:75]: OCD 🗡️  (not need to fix it, only if you want)

File: packages/adminpanel-v2/src/stores/CustomerStore.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	const getActiveCustomer = () => {
+		const store = get(customerStore)
+		return store.customers.find((c) => c.id == store.active.customer)
+	}
+
+	const getActiveProject = () => {
+		const store = get(customerStore)
+		return store.customers.find((c) => c.id == store.active.customer)?.projects.find((p) => p.id == store.active.project)
```
Comment (stormboat) [inline packages/adminpanel-v2/src/stores/CustomerStore.ts:63]: are there any requirements on line length on javascript or svelte? (ref python 79 characters line length in PEP8)

File: packages/adminpanel-v2/src/components/customerpage/customerlist/customers/Customers.svelte [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
+	import { customers } from "../../../../stores/CustomerStore";
     import Customer from "./customer/Customer.svelte";
-    export let customers : TCustomer[]
     export let selectCustomer : Function
 </script>
 
 <div class="left-component">
     <ul class="customer-list">
-        {#each customers as customer}
+        {#each $customers.customers as customer}
```
Comment (stormboat) [inline packages/adminpanel-v2/src/components/customerpage/customerlist/customers/Customers.svelte:10]: What does it mean when you use it inline inside a loop, does it have any specific cases then or do you always call it with the $ sign when you use a reactive variable?

File: packages/adminpanel-v2/src/components/customerpage/customerdetails/activities/Activities.svelte
Comment (yvindo) [reply packages/adminpanel-v2/src/components/customerpage/customerdetails/activities/Activities.svelte:1]: $: creates a responsive variable, which listens to changes in variables/functions used in its definition.

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 1 above)
Comment (yvindo) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:37]: The customer store is currently only using mock data, so the store can be played around with. Changes will not be stored in any way between sessions until #716 is completed.

Comment (yvindo) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:37]: The id here refers to the id defined in the customer type used in the mock data. Once the store is connected to the api, the object structure will have to be updated to reflect the data fetched from the backend.

File: packages/adminpanel-v2/src/components/customerpage/customerdetails/activities/Activities.svelte
Comment (stormboat) [reply packages/adminpanel-v2/src/components/customerpage/customerdetails/activities/Activities.svelte:1]: thanks

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 2 above)
Comment (yvindo) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:45]: The store is responsible for maintaining the state of the frontend. Once connected to the backend, async functions must be added to update the data in the backend, wait for a response, and then update the frontend store. Such functions will be added directly in the store, so backend connection is not available without going through the store.

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 3 above)
Comment (yvindo) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:56]: Correct

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 1 above)
Comment (stormboat) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:37]: That makes sense now, thanks

Comment (stormboat) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:37]: Thank you! 🥇

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 4 above)
Comment (yvindo) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:75]: The indenting is correct in my local branch, don´t know why git does not notice such differences.

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 5 above)
Comment (yvindo) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:63]: Not that I am aware of. However, we can define team preferences for when to split longer functions into several lines.

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 2 above)
Comment (stormboat) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:45]: Thanks for a good a extensive explanation!

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 4 above)
Comment (stormboat) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:75]: aah okay good

File: packages/adminpanel-v2/src/stores/CustomerStore.ts (hunk 5 above)
Comment (stormboat) [reply packages/adminpanel-v2/src/stores/CustomerStore.ts:63]: Yeah okay, nice!

File: packages/adminpanel-v2/src/components/customerpage/customerlist/customers/Customers.svelte (hunk 6 above)
Comment (yvindo) [reply packages/adminpanel-v2/src/components/customerpage/customerlist/customers/Customers.svelte:10]: In this case $customers refers to the store. The $ sign is generally reserved to store referencing, except for when used to declare reactive variables. When referring to a store, $ must be used for store variables, but not when referring to store functions.

Comment (stormboat) [reply packages/adminpanel-v2/src/components/customerpage/customerlist/customers/Customers.svelte:10]: aaaah okay, nice. That makes sense too

**Review by stormboat — APPROVED** [review]

**Pull request merged**